File: elixir/serviceradar_core/test/serviceradar/edge/lower_bound_corpus_test.exs
defmodule ServiceRadar.Edge.LowerBoundCorpusTest do
  @moduledoc """
  Task 1.5-h: the SHARED LOWER-BOUND CORPUS, this runtime's half.

  ITS OWN INVENTORY, NOT AN EXTENSION OF THE COUNT CORPUS. The upper table proves each site's
  ceiling with an N/N+1 pair; extending it mechanically with one empty row per site would
  assert independent proofs the implementations do not provide, because several local emptiness
  predicates are SHADOWED by a later check that refuses the same input.

  This file asks a different question and answers it per runtime: what does the BOUNDARY do
  with an empty collection, and what would removing the LOCAL arm actually change? The second
  answer is the `ex_removal` column, and it is MEASURED. See the manifest for what each token
  licenses a row to claim; two matter here:

    * `combined` -- this runtime has NO separable zero arm at that site. The raw plan gate is a
      single conjunction over the declared count, the ceiling and the supplied length, so there
      is nothing to remove independently and the row claims nothing about it.
    * `silent` -- the boundary still refuses under the SAME reason because a later gate catches
      it, so no verdict-based row can kill the removal. The row records boundary behaviour.

  TWO CONTROLS PER RULE. Zero refused alone does not pin a minimum of 1: an implementation
  demanding two elements refuses zero exactly as before. The ONE-element acceptance is what
  makes it a bound rather than a prohibition.
  """
  use ExUnit.Case, async: true

  alias ServiceRadar.Edge.AssignmentValidate
  alias ServiceRadar.Edge.HashGrammar
  alias ServiceRadar.Edge.PlanValidate
  alias ServiceRadar.Edge.RecoveryValidate
  alias Serviceradar.Edge.V1.EdgeLossManifestPageV1
  alias Serviceradar.Edge.V1.ScheduledPlanHeaderV1
  alias Serviceradar.Edge.V1.ScheduledPlanPageV1
  alias Serviceradar.Edge.V1.SweepAssignmentRecordV1

  describe "the manifest and this runtime agree" do
    test "the inventory matches the manifest on EVERY column, both directions" do
      # EVERY COLUMN, both runtimes' removal classes included. Omitting `go_removal` would let
      # the other runtime's measured claim drift with nothing here noticing, and omitting
      # at/over would let the declared-scalar ceiling move silently.
      inventory = %{
        "plan_pages_decoded" => {"collection", "retags", "retags", "n/a", "n/a", "-"},
        "plan_pages_raw" => {"collection", "silent", "combined", "n/a", "n/a", "-"},
        "plan_ranges" => {"collection", "admits", "admits", "n/a", "n/a", "-"},
        "recovery_pages_decoded" => {"collection", "crashes", "crashes", "n/a", "n/a", "-"},
        "recovery_pages_raw" => {"collection", "silent", "silent", "n/a", "n/a", "-"},
        "recovery_spans_chain" => {"collection", "retags", "retags", "n/a", "n/a", "-"},
        "recovery_spans_single" => {"collection", "retags", "n/a", "n/a", "n/a", "1.6-d"},
        "tombstone_declared_count" => {"scalar", "admits", "n/a", "1024", "1025", "1.6-d"}
      }

      removals = ["admits", "crashes", "retags", "silent", "combined", "n/a"]
      rows = corpus()

      assert length(rows) == map_size(inventory), "the manifest has a duplicate or missing row"
      assert MapSet.new(rows, & &1.site) == MapSet.new(Map.keys(inventory))

      for r <- rows do
        assert {r.kind, r.go_removal, r.ex_removal, r.at, r.over, r.owner} == inventory[r.site],
               "#{r.site}: the manifest and the inventory disagree"

        # Only the DECLARED-SCALAR rule freezes a ceiling here; a collection's ceiling belongs
        # to the count corpus, and restating it would be a second inventory to keep in sync.
        assert r.at != "n/a" == (r.kind == "scalar"),
               "#{r.site}: kind=#{r.kind} but at=#{r.at}"

        assert r.at == "n/a" == (r.over == "n/a"),
               "#{r.site}: at and over must both be set or both be n/a"

        assert r.ex_removal in removals and r.go_removal in removals,
               "#{r.site}: removal tokens #{r.go_removal}/#{r.ex_removal} are not both known"

        assert {r.zero, r.one} == {"refuse", "accept"},
               "#{r.site}: a lower bound of 1 is zero=refuse/one=accept"

        assert r.ex_removal == "n/a" == (r.owner != "-"),
               "#{r.site}: an absent peer must name an owner and a present one must not"
      end
    end
  end

  describe "plan collections" do
    test "plan_pages_decoded and plan_pages_raw" do
      raw = row("plan_pages_raw")
      {h, pages} = plan()

      assert {:error, :page_bounds} = PlanValidate.validate(h, [])
      # ONE PAGE IS ADMITTED -- without it a validator demanding two would satisfy the row.
      assert {:ok, _} = PlanValidate.validate(h, pages)

      header_bytes = ScheduledPlanHeaderV1.encode(h)
      raw_pages = Enum.map(pages, &ScheduledPlanPageV1.encode/1)

      assert {:error, :page_bounds} =
               AssignmentValidate.validate_bytes_against_plan_bytes(<<>>, header_bytes, [])

      # A COMBINED GATE, so this row records the BOUNDARY only. `bound_page_count/2` is one
      # conjunction over the declared count, the ceiling and the supplied length -- there is no
      # zero arm to remove on its own, and the manifest says so rather than implying a proof.
      assert raw.ex_removal == "combined"

      # ONE PAGE ACCEPTED THROUGH THE PUBLIC BOUNDARY, with a VALID assignment bound to this
      # plan. `validate_bytes_against_plan_bytes/3` validates the record and its plan relation
      # after the count gate, so an empty record can only ever yield an error -- and a row
      # asserting `{:error, _}` stays green with the minimum tightened to two pages, which is
      # exactly the drift the acceptance exists to catch.
      record = SweepAssignmentRecordV1.encode(assignment_for(h, pages))

      assert {:ok, _} =
               AssignmentValidate.validate_bytes_against_plan_bytes(
                 record,
                 header_bytes,
                 raw_pages
               )
    end

    test "plan_ranges -- the load-bearing plan arm" do
      r = row("plan_ranges")
      assert r.ex_removal == "admits"

      {h, pages} = plan()
      assert {:ok, _} = PlanValidate.validate(h, pages)

      # RESEALED so the empty page is refused by its OWN rule and not by a stale digest,
      # declared total or commitment.
      page = %{hd(pages) | ranges: []}
      page = %{page | page_sha256: HashGrammar.plan_page_digest(page)}
      {:ok, commitment} = HashGrammar.plan_mtr_ordinal_range_commitment([page])

      eh = %{
        h
        | plan_root_sha256: HashGrammar.plan_root([page]),
          total_target_count: 0,
          mtr_ordinal_range_commitment: commitment
      }

      eh = %{eh | execution_plan_sha256: HashGrammar.plan_header_digest(eh)}

      assert {:error, :page_bounds} = PlanValidate.validate(eh, [page])
    end
  end

  describe "recovery collections" do
    test "recovery_pages_decoded and recovery_pages_raw" do
      decoded = row("recovery_pages_decoded")
      pages = [manifest_page()]
      root = HashGrammar.manifest_root(pages)

      assert {:error, :manifest_empty} = RecoveryValidate.manifest_chain([], root)
      assert :ok = RecoveryValidate.manifest_chain(pages, root)

      raw = Enum.map(pages, &EdgeLossManifestPageV1.encode/1)
      assert {:error, :manifest_empty} = RecoveryValidate.manifest_chain_from_raw([], root)
      assert :ok = RecoveryValidate.manifest_chain_from_raw(raw, root)

      # THE DECODED ARM IS NOT MERELY A VERDICT here either: removing its clause leaves the
      # remaining head matching an empty list and raising, so it is what keeps a public
      # validator from crashing on attacker-supplied input.
      assert decoded.ex_removal == "crashes"
    end

    test "recovery_spans_chain" do
      r = row("recovery_spans_chain")
      pages = [manifest_page()]

      assert :ok = RecoveryValidate.manifest_chain(pages, HashGrammar.manifest_root(pages))

      empty = %{hd(pages) | classification_spans: []}
      empty = %{empty | page_sha256: HashGrammar.manifest_page_digest(empty)}

      # THE EXACT REFUSAL REASON, not merely "refused": the arm is `retags`, so with the local
      # check removed the boundary still refuses under the span-body reason. Asserting the
      # reason is what makes this row kill that removal.
      assert r.ex_removal == "retags"

      assert {:error, :manifest_bounds} =
               RecoveryValidate.manifest_chain([empty], HashGrammar.manifest_root([empty]))
    end
  end

  describe "sites with no peer here" do
    test "the single-page span site and the signed declared count are 1.6-d's" do
      for site <- ["recovery_spans_single", "tombstone_declared_count"] do
        r = row(site)
        assert r.ex_removal == "n/a"
        assert r.owner == "1.6-d"
      end
    end
  end

  # ---------------------------------------------------------------------------
  # adapters
  # ---------------------------------------------------------------------------

  defp plan do
    h = "plan_header.bin" |> fixture() |> ScheduledPlanHeaderV1.decode()
    p = "plan_page.bin" |> fixture() |> ScheduledPlanPageV1.decode()

    # ZERO MTR, stated explicitly and RESEALED. The committed range admits two ordinals while
    # the committed assignment expects none, so an accepted control built from both unchanged
    # is refused for the MTR RELATION -- another rule's outcome standing in for this one's.
    ranges =
      Enum.map(p.ranges, fn r ->
        r = %{r | mtr_ordinal_count: 0, mtr_admission_budget: 0}
        %{r | range_sha256: HashGrammar.range_digest(r)}
      end)

    page = %{p | page_index: 0, page_count: 1, prev_page_sha256: <<>>, ranges: ranges}
    page = %{page | page_sha256: HashGrammar.plan_page_digest(page)}
    {:ok, commitment} = HashGrammar.plan_mtr_ordinal_range_commitment([page])

    h = %{
      h
      | page_count: 1,
        plan_root_sha256: HashGrammar.plan_root([page]),
        total_target_count: Enum.sum(Enum.map(page.ranges, & &1.target_count)),
        mtr_ordinal_range_commitment: commitment
    }

    {%{h | execution_plan_sha256: HashGrammar.plan_header_digest(h)}, [page]}
  end

  # EXACTLY ONE SPAN. The committed fixture carries THREE, so using it unchanged would have
  # proved that a three-span page is admitted -- true, and not the claim. A minimum of 1 is
  # pinned only by admitting a page carrying exactly one. The page digest is resealed after the
  # reduction, and every caller recomputes the manifest root from the page it actually built.
  # A VALID assignment BOUND TO THIS PLAN: only the fields that bind a record to a plan are
  # re-pointed, so the accepted control differs from the refused one in page count alone.
  defp assignment_for(header, pages) do
    range = hd(hd(pages).ranges)
    base = "assignment_zero_mtr.bin" |> fixture() |> SweepAssignmentRecordV1.decode()

    %{
      base
      | execution_plan_id: header.execution_plan_id,
        execution_plan_sha256: header.execution_plan_sha256,
        network_scope_id: header.network_scope_id,
        target_range_id: range.range_id,
        target_range_sha256: range.range_sha256,
        availability_policy_id: header.availability_policy_id,
        check_set_sha256: header.check_set_sha256
    }
  end

  defp manifest_page do
    p = "manifest_page.bin" |> fixture() |> EdgeLossManifestPageV1.decode()

    page = %{
      p
      | page_index: 0,
        page_count: 1,
        prev_page_sha256: <<>>,
        terminal: true,
        classification_spans: Enum.take(p.classification_spans, 1)
    }

    %{page | page_sha256: HashGrammar.manifest_page_digest(page)}
  end

  # ---------------------------------------------------------------------------
  # manifest
  # ---------------------------------------------------------------------------

  defp row(site),
    do: Enum.find(corpus(), &(&1.site == site)) || flunk("no lower-bound row for #{site}")

  defp corpus do
    corpus_path()
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.reject(&String.starts_with?(String.trim(&1), "#"))
    |> Enum.map(fn line ->
      case String.split(String.trim(line), ~r/\s+/) do
        [site, kind, zero, one, at, over, go_removal, ex_removal, owner] ->
          %{
            site: site,
            kind: kind,
            zero: zero,
            one: one,
            at: at,
            over: over,
            go_removal: go_removal,
            ex_removal: ex_removal,
            owner: owner
          }

        other ->
          flunk("lower-bound row #{inspect(other)} does not have 9 fields")
      end
    end)
  end

  defp corpus_path,
    do: Path.expand("../../../../../proto/edge/v1/testdata/lower_bound_corpus.txt", __DIR__)

  defp fixture(name), do: corpus_path() |> Path.dirname() |> Path.join(name) |> File.read!()
end

File: go/pkg/edge/edgerecord/lower_bound_corpus_test.go
package edgerecord

import (
	"bufio"
	"errors"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"

	edgev1 "github.com/carverauto/serviceradar/proto/edge/v1"
	"google.golang.org/protobuf/proto"
)

// Task 1.5-h: THE SHARED LOWER-BOUND CORPUS.
//
// ITS OWN INVENTORY, NOT AN EXTENSION OF THE COUNT CORPUS. The upper table proves each site's
// ceiling with an N/N+1 pair; extending it mechanically with one empty row per site would
// assert independent proofs the implementations do not provide, because several local
// emptiness predicates are SHADOWED by a later check that refuses the same input.
//
// So this file asks a different question and answers it per runtime: what does the BOUNDARY
// do with an empty collection, and what would removing the LOCAL arm actually change? The
// second answer is the `go_removal` column, and it is MEASURED -- see the manifest for what
// each token licenses a row to claim.
//
// TWO CONTROLS PER RULE. Zero refused alone does not pin a minimum of 1: an implementation
// demanding two elements refuses zero exactly as before. The ONE-element acceptance is what
// makes it a bound rather than a prohibition.

type lowerRow struct {
	site, kind           string
	zero, one            string
	at, over             int
	hasBounds            bool
	goRemoval, exRemoval string
	owner                string
}

func lowerCorpus(t *testing.T) []lowerRow {
	t.Helper()

	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}

	f, err := os.Open(filepath.Join(wd, "..", "..", "..", "..", "proto", "edge", "v1", "testdata", "lower_bound_corpus.txt"))
	if err != nil {
		t.Fatalf("open lower-bound corpus: %v", err)
	}
	defer func() { _ = f.Close() }()

	var out []lowerRow

	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		fields := strings.Fields(line)
		if len(fields) != 9 {
			t.Fatalf("lower-bound row %q has %d fields, want 9", line, len(fields))
		}

		row := lowerRow{
			site: fields[0], kind: fields[1], zero: fields[2], one: fields[3],
			goRemoval: fields[6], exRemoval: fields[7], owner: fields[8],
		}

		if fields[4] == verdictNA && fields[5] != verdictNA {
			t.Fatalf("row %q: at is n/a but over is %q -- one would be ignored", line, fields[5])
		}

		if fields[4] != verdictNA {
			at, err := strconv.Atoi(fields[4])
			if err != nil {
				t.Fatalf("row %q: at: %v", line, err)
			}

			over, err := strconv.Atoi(fields[5])
			if err != nil {
				t.Fatalf("row %q: over: %v", line, err)
			}

			row.at, row.over, row.hasBounds = at, over, true
		}

		out = append(out, row)
	}

	if err := sc.Err(); err != nil {
		t.Fatalf("scan lower-bound corpus: %v", err)
	}

	return out
}

func lowerRowFor(t *testing.T, site string) lowerRow {
	t.Helper()

	for _, r := range lowerCorpus(t) {
		if r.site == site {
			return r
		}
	}

	t.Fatalf("no lower-bound row for site %q", site)

	return lowerRow{}
}

// TestLowerCorpusInventory pins the full row. The `*_removal` tokens are the part most likely
// to drift optimistically -- an arm that becomes shadowed still passes every verdict row, so
// nothing but this comparison notices the claim is now too strong.
func TestLowerCorpusInventory(t *testing.T) {
	want := map[string]lowerRow{
		"plan_pages_decoded":       {kind: "collection", goRemoval: "retags", exRemoval: "retags", owner: "-"},
		"plan_pages_raw":           {kind: "collection", goRemoval: "silent", exRemoval: "combined", owner: "-"},
		"plan_ranges":              {kind: "collection", goRemoval: "admits", exRemoval: "admits", owner: "-"},
		"recovery_pages_decoded":   {kind: "collection", goRemoval: "crashes", exRemoval: "crashes", owner: "-"},
		"recovery_pages_raw":       {kind: "collection", goRemoval: "silent", exRemoval: "silent", owner: "-"},
		"recovery_spans_chain":     {kind: "collection", goRemoval: "retags", exRemoval: "retags", owner: "-"},
		"recovery_spans_single":    {kind: "collection", goRemoval: "retags", exRemoval: verdictNA, owner: proofGroup16D},
		"tombstone_declared_count": {kind: "scalar", goRemoval: "admits", exRemoval: verdictNA, owner: proofGroup16D, at: 1024, over: 1025, hasBounds: true},
	}

	removals := map[string]bool{
		"admits": true, "crashes": true, "retags": true,
		"silent": true, "combined": true, verdictNA: true,
	}

	rows := lowerCorpus(t)
	if len(rows) != len(want) {
		t.Fatalf("corpus has %d rows, inventory has %d -- a duplicate or missing row", len(rows), len(want))
	}

	// A LENGTH CHECK ALONE ADMITS SUBSTITUTION: duplicating one site while deleting another
	// keeps the count, and every per-row check below then passes twice on the survivor.
	seen := map[string]bool{}

	for _, r := range rows {
		if seen[r.site] {
			t.Fatalf("%s appears twice in the corpus", r.site)
		}

		seen[r.site] = true
	}

	for _, r := range rows {
		w, ok := want[r.site]
		if !ok {
			t.Fatalf("corpus row %q is not in the inventory", r.site)
		}

		if r.kind != w.kind || r.goRemoval != w.goRemoval || r.exRemoval != w.exRemoval ||
			r.owner != w.owner || r.at != w.at || r.over != w.over || r.hasBounds != w.hasBounds {
			t.Fatalf("%s: manifest and inventory disagree", r.site)
		}

		if r.zero != verdictRefuse || r.one != verdictAccept {
			t.Fatalf("%s: a lower bound of 1 is zero=refuse/one=accept, got %s/%s", r.site, r.zero, r.one)
		}

		if !removals[r.goRemoval] || !removals[r.exRemoval] {
			t.Fatalf("%s: removal tokens %s/%s are not both recognised", r.site, r.goRemoval, r.exRemoval)
		}

		// An absent peer must NAME AN OWNER. `n/a` is the only token that may.
		if (r.exRemoval == verdictNA) != (r.owner != "-") {
			t.Fatalf("%s: ex_removal=%s owner=%s -- an absent peer must name an owner and a present one must not",
				r.site, r.exRemoval, r.owner)
		}

		// Only the DECLARED-SCALAR rule freezes a ceiling here; a collection's ceiling belongs
		// to the count corpus and restating it would create a second inventory to keep in sync.
		if r.hasBounds != (r.kind == "scalar") {
			t.Fatalf("%s: kind=%s but hasBounds=%t", r.site, r.kind, r.hasBounds)
		}

		if r.hasBounds && r.over != r.at+1 {
			t.Fatalf("%s: %d/%d is not an adjacent pair", r.site, r.at, r.over)
		}

		// The reader only inspects `over` when `at` is numeric, so a row reading `n/a 1025`
		// would otherwise carry a silently ignored value.
		if !r.hasBounds && r.over != 0 {
			t.Fatalf("%s: at is n/a but over parsed as %d", r.site, r.over)
		}
	}
}

// ---------------------------------------------------------------------------
// collection sites -- zero refused, one accepted, at the BOUNDARY
// ---------------------------------------------------------------------------

func TestLowerSitePlanPages(t *testing.T) {
	decoded := lowerRowFor(t, "plan_pages_decoded")
	raw := lowerRowFor(t, "plan_pages_raw")

	h, pages := buildPlan(t, mustUUID(t), d32(0x81), [][]uint64{{1}})

	if err := ValidatePlanPages(h, nil); !errors.Is(err, ErrPlanBounds) {
		t.Fatalf("an empty decoded page list = %v, want ErrPlanBounds", err)
	}

	// ONE PAGE IS ADMITTED. Without it the row would be satisfied by a validator demanding two.
	if err := ValidatePlanPages(h, pages); err != nil {
		t.Fatalf("a one-page plan must be accepted: %v", err)
	}

	if _, _, err := ValidatePlanFromRaw(headerBytes(t, h), [][]byte{}); !errors.Is(err, ErrPlanBounds) {
		t.Fatalf("an empty raw page list = %v, want ErrPlanBounds", err)
	}

	if _, _, err := ValidatePlanFromRaw(headerBytes(t, h), rawPlanPagesOf(t, pages)); err != nil {
		t.Fatalf("a one-page raw plan must be accepted: %v", err)
	}

	// The raw row records BOUNDARY behaviour only: its local arm is `silent`, so removing it
	// changes no verdict here -- the decoded gate refuses the same input for the same reason.
	if raw.goRemoval != "silent" || decoded.goRemoval != "retags" {
		t.Fatalf("removal classes changed: raw=%s decoded=%s", raw.goRemoval, decoded.goRemoval)
	}
}

func TestLowerSitePlanRanges(t *testing.T) {
	r := lowerRowFor(t, "plan_ranges")

	if r.goRemoval != "admits" {
		t.Fatalf("plan_ranges is the load-bearing plan arm; manifest says %s", r.goRemoval)
	}

	h, pages := buildPlan(t, mustUUID(t), d32(0x82), [][]uint64{{1}})

	if err := ValidatePlanPages(h, pages); err != nil {
		t.Fatalf("a one-range page must be accepted: %v", err)
	}

	empty, ok := proto.Clone(pages[0]).(*edgev1.ScheduledPlanPageV1)
	if !ok {
		t.Fatal("clone plan page")
	}

	empty.Ranges = nil
	empty.PageSha256 = PlanPageDigest(empty)

	one := []*edgev1.ScheduledPlanPageV1{empty}

	eh, ok := proto.Clone(h).(*edgev1.ScheduledPlanHeaderV1)
	if !ok {
		t.Fatal("clone plan header")
	}

	// RESEALED so the empty page is refused by its own rule and not by a stale digest,
	// declared total or commitment.
	eh.PlanRootSha256 = PlanRoot(one)
	eh.TotalTargetCount = 0
	eh.MtrOrdinalRangeCommitment = mustPlanCommitment(t, one)
	eh.ExecutionPlanSha256 = PlanHeaderDigest(eh)

	if err := ValidatePlanPages(eh, one); !errors.Is(err, ErrPlanBounds) {
		t.Fatalf("a page with no ranges = %v, want ErrPlanBounds", err)
	}
}

func TestLowerSiteRecoveryPages(t *testing.T) {
	decoded := lowerRowFor(t, "recovery_pages_decoded")
	raw := lowerRowFor(t, "recovery_pages_raw")

	pages := buildManifest(t, mustUUID(t), [][]*edgev1.EdgeClassificationSpanV1{spansOf(t, 1)})
	root := ManifestRoot(pages)

	if err := ValidateManifestChain(nil, root); !errors.Is(err, ErrManifestEmpty) {
		t.Fatalf("an empty decoded manifest = %v, want ErrManifestEmpty", err)
	}

	if err := ValidateManifestChain(pages, root); err != nil {
		t.Fatalf("a one-page manifest must be accepted: %v", err)
	}

	if err := ValidateManifestChainFromRaw([][]byte{}, root); !errors.Is(err, ErrManifestEmpty) {
		t.Fatalf("an empty raw manifest = %v, want ErrManifestEmpty", err)
	}

	if err := ValidateManifestChainFromRaw(rawPagesOf(t, pages), root); err != nil {
		t.Fatalf("a one-page raw manifest must be accepted: %v", err)
	}

	// THE DECODED ARM IS NOT MERELY A VERDICT. Removing it indexes page zero of an empty
	// slice, so it is what keeps a public validator from PANICKING on attacker-supplied
	// input -- which is why the manifest classes it `crashes` and not `retags`.
	if decoded.goRemoval != "crashes" || raw.goRemoval != "silent" {
		t.Fatalf("removal classes changed: decoded=%s raw=%s", decoded.goRemoval, raw.goRemoval)
	}
}

func TestLowerSiteRecoverySpans(t *testing.T) {
	chain := lowerRowFor(t, "recovery_spans_chain")
	single := lowerRowFor(t, "recovery_spans_single")

	if single.exRemoval != verdictNA || single.owner != proofGroup16D {
		t.Fatalf("the single-page span site is Go-only; manifest says %s/%s", single.exRemoval, single.owner)
	}

	pages := buildManifest(t, mustUUID(t), [][]*edgev1.EdgeClassificationSpanV1{spansOf(t, 1)})

	if err := ValidateManifestChain(pages, ManifestRoot(pages)); err != nil {
		t.Fatalf("a one-span page must be accepted: %v", err)
	}

	// THE SINGLE-PAGE SITE IS DRIVEN THROUGH ITS PUBLIC BOUNDARY, as the upper pair is.
	// `validateSingleManifestPage` is private and reachable only via `ValidateRecoveryControl`,
	// so calling it directly would prove the predicate while leaving the ROUTE to it untested
	// -- and the route is half of what the site is.
	wrap := func(spans []*edgev1.EdgeClassificationSpanV1) (*edgev1.EdgeRecordV1, AuthorizationPolicy) {
		page := buildManifest(t, mustUUID(t),
			[][]*edgev1.EdgeClassificationSpanV1{spansOf(t, 1)})[0]

		// Finished FIRST -- spans, then count/index/terminal, then its own digest -- and only
		// then is the scope digest taken and the wrapper built around it.
		page.ClassificationSpans = spans
		page.PageIndex, page.PageCount, page.Terminal, page.PrevPageSha256 = 0, 1, true, nil
		page.PageSha256 = ManifestPageDigest(page)

		body := &edgev1.EdgeRecoveryControlPayloadV1{
			Body: &edgev1.EdgeRecoveryControlPayloadV1_ManifestPage{ManifestPage: page},
		}

		return scopeControlRecord(t, body, page.GetRecoveryId(), ManifestPageScopeDigest(page))
	}

	checkSingle := func(name string, spans []*edgev1.EdgeClassificationSpanV1, want error) {
		t.Helper()

		rec, policy := wrap(spans)

		// SIGNED-VALID BEFORE ANY PRODUCTION CLAIM, or a stale signature could masquerade as
		// bound evidence.
		if err := ValidateRecordSigned(rec, policy); err != nil {
			t.Fatalf("%s: the wrapper must pass the signed boundary first: %v", name, err)
		}

		err := ValidateRecoveryControl(rec, rec.GetOutputContract(), policy)
		if want == nil {
			if err != nil {
				t.Fatalf("%s: must be accepted: %v", name, err)
			}

			return
		}

		if !errors.Is(err, want) {
			t.Fatalf("%s: = %v, want %v", name, err, want)
		}
	}

	checkSingle("one span", spansOf(t, 1), nil)

	empty, ok := proto.Clone(pages[0]).(*edgev1.EdgeLossManifestPageV1)
	if !ok {
		t.Fatal("clone manifest page")
	}

	empty.ClassificationSpans = nil
	empty.PageSha256 = ManifestPageDigest(empty)

	one := []*edgev1.EdgeLossManifestPageV1{empty}

	// THE EXACT REFUSAL REASON, not merely "refused". Both arms are `retags`: with the local
	// check removed the boundary still refuses, under the span-body reason instead. Asserting
	// the reason is what makes these rows kill that removal.
	if err := ValidateManifestChain(one, ManifestRoot(one)); !errors.Is(err, ErrManifestBounds) {
		t.Fatalf("a page with no spans = %v, want ErrManifestBounds", err)
	}

	checkSingle("no spans", nil, ErrManifestBounds)

	if chain.goRemoval != "retags" {
		t.Fatalf("the chain span arm is classed %s", chain.goRemoval)
	}
}

// ---------------------------------------------------------------------------
// the DECLARED SCALAR -- four controls on the signed path
// ---------------------------------------------------------------------------

// TestLowerSiteTombstoneDeclaredCount covers the rule this task freezes: a signed tombstone's
// `manifest_page_count` is bounded 1..MaxManifestPages, not merely non-zero.
//
// IT IS A DIFFERENT SITE FROM THE PAGE-LIST BOUND. `ValidateTombstone` reconciles a
// declaration against supplied pages; this path receives NO pages, so nothing questions the
// declaration and the scope digest commits it exactly as signed.
//
// EACH CONTROL IS FULLY RESCOPED AND RE-SIGNED. The tombstone is built first, its scope digest
// taken from the finished body, and the payload, source claim, envelope and both signatures
// built around it -- so a control never reuses a wrapper whose signature describes a different
// declaration. ValidateRecordSigned is asserted BEFORE the production claim, or a stale
// signature could masquerade as bound evidence.
func TestLowerSiteTombstoneDeclaredCount(t *testing.T) {
	r := lowerRowFor(t, "tombstone_declared_count")

	if r.kind != "scalar" || r.owner != proofGroup16D {
		t.Fatalf("the declared count is a Go-only scalar rule owned by 1.6-d; manifest says %s/%s", r.kind, r.owner)
	}

	wrap := func(count uint32) (*edgev1.EdgeRecordV1, AuthorizationPolicy) {
		id := mustUUID(t)

		tomb := &edgev1.SpoolLossTombstoneV1{
			RecoveryId:         id,
			PriorSpoolId:       mustUUID(t),
			NewSpoolId:         mustUUID(t),
			DetectedAtUnixNano: 1_700_000_000_000_000_000,
			DigestVersion:      RecoveryDigestVersion,
			ManifestPageCount:  count,
			ManifestRootSha256: d32(0x83),
			Reason:             "corpus",
		}

		body := &edgev1.EdgeRecoveryControlPayloadV1{
			Body: &edgev1.EdgeRecoveryControlPayloadV1_Tombstone{Tombstone: tomb},
		}

		return scopeControlRecord(t, body, id, TombstoneScopeDigest(tomb))
	}

	check := func(name string, count uint32, want error) {
		t.Helper()

		rec, policy := wrap(count)

		if err := ValidateRecordSigned(rec, policy); err != nil {
			t.Fatalf("%s: the wrapper must pass the signed boundary first: %v", name, err)
		}

		err := ValidateRecoveryControl(rec, rec.GetOutputContract(), policy)
		if want == nil {
			if err != nil {
				t.Fatalf("%s: count %d must be accepted: %v", name, count, err)
			}

			return
		}

		if !errors.Is(err, want) {
			t.Fatalf("%s: count %d = %v, want %v", name, count, err, want)
		}
	}

	check("zero", 0, ErrTombstoneMismatch)
	// The frozen minimum, ACCEPTED -- without it a gate demanding two pages passes "zero".
	check("minimum", 1, nil)
	check("at ceiling", uint32(r.at), nil)
	check("one over", uint32(r.over), ErrTombstoneMismatch)
}
